massifs/bloberrors.go

```diff
@@ -0,0 +1,129 @@
+package massifs
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	azStorageBlob "github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+)
+
+const (
+	azblobBlobNotFound = "BlobNotFound"
+)
+
+func AsStorageError(err error) (azStorageBlob.StorageError, bool) {
+	serr := &azStorageBlob.StorageError{}
+	//nolint
+	ierr, ok := err.(*azStorageBlob.InternalError)
+	if ierr == nil || !ok {
+		return azStorageBlob.StorageError{}, false
+	}
+	if !ierr.As(&serr) {
+		return azStorageBlob.StorageError{}, false
+	}
+	return *serr, true
+}
+
+func AsResponseError(err error) (azcore.ResponseError, bool) {
+
+	var ok bool
+	var rerr *azcore.ResponseError
+
+	//nolint
+	if rerr, ok = err.(*azcore.ResponseError); ok {
+		return *rerr, true
+	}
+
+	// check for an InternalError that has ResponseError as its cause
+	rerr = &azcore.ResponseError{}
+
+	//nolint
+	ierr, ok := err.(*azStorageBlob.InternalError)
+	if ierr == nil || !ok {
+		return azcore.ResponseError{}, false
+	}
+	if !ierr.As(&rerr) {
+		return azcore.ResponseError{}, false
+	}
+	return *rerr, true
+}
+
+// WrapBlobNotFound tranlsates the err to ErrBlobNotFound if the actual error is
+// the azure sdk blob not found error. In all cases where the original err is
+// not the azure BlobNot found, the original err is returned as is. Including
+// the case where it is nil
+func WrapBlobNotFound(err error) error {
+	if err == nil {
+		return nil
+	}
+	serr, ok := AsStorageError(err)
+	if !ok {
+		return err
+	}
+	if serr.ErrorCode != azblobBlobNotFound {
+		return err
+	}
+	return fmt.Errorf("%s: %w", err.Error(), ErrBlobNotFound)
+}
+
+func IsBlobNotFound(err error) bool {
+	if err == nil {
+		return false
+	}
+	if errors.Is(err, ErrBlobNotFound) {
+		return true
+	}
+	serr, ok := AsStorageError(err)
+	if !ok {
+		return false
+	}
+	if serr.ErrorCode != azblobBlobNotFound {
+		return false
+	}
+	return true
+}
+
+// IsRateLimiting detects if the error is HTTP Status 429 Too Many Requests
+// The recomended wait time is returned if it is available. If the returned wait
+// time is zero, the caller should apply an appropriate default backoff.
+func IsRateLimiting(err error) (time.Duration, bool) {
```

Comment:
is it worth making an `ErrRateLimiting`? and using the errors interfaces?

Reply:
for example you could do:

```
var rateLimitingErr *ErrRateLimiting
if errors.As(err, &rateLimitingErr) {
	// return some useful message about rate limiting
        return fmt.Sprintf("rate limit reached, retry after: %v", rateLimitingErr.WaitTime)
}
```

Reply:
Definitely not, that forced check return and convert on the caller.

Reply:
resolved with as it works and is tested and @robinbryce has raised a td ticket: [User Story 9917](https://dev.azure.com/jitsuin/avid/_workitems/edit/9917): TD: IsBlobError and IsRateLimiting helpers